samples/simulation/qaoa/QAOA.qs

```diff
@@ -0,0 +1,240 @@
+    /// to a parameter J_ij between qubit states i and j.
+    function createHamiltonianCouplings(penalty: Double) : Double[] {
+        // Calculate Hamiltonian coupling parameters based on the given costs and penalty
+        mutable coupling = new Double[36];
+
+        // Most elements of J_ij equal 2*penalty, so set all elements to this value, 
+        // then overwrite the exceptions
+        for (i in 0..35)
+        {
+            set coupling w/= i <- 2.0 * penalty;
+        }
+        set coupling w/= 2 <- penalty;
+        set coupling w/= 9 <- penalty;
+        set coupling w/= 29 <- penalty;
+
+        return coupling;
+    }
+    
+    /// # Summary
+    /// Perform the QAOA algorithm for this Ising Hamiltonian
+    ///
+    /// # Input
+    /// ## weights
+    /// Instance Hamiltonian parameters or "weights" as an array where each element corresponds to a 
+    /// parameter h_j for qubit state j.
+    /// ## couplings
+    /// Instance Hamiltonian coupling parameters as an array, where each element corresponds
+    /// to a parameter J_ij between qubit states i and j.
+    /// ## timeX
+    /// Time evolution for PauliX operations
+    /// ## timeZ
+    /// Time evolution for PauliX operations
+    operation PerformQAOA(weights : Double[], couplings : Double[], timeX : Double[], timeZ : Double[]) : Bool[] {
+        EqualityFactI(Length(timeX), Length(timeZ), "timeZ and timeX are not the same length");
+
+        // Run the QAOA circuit
+        mutable result = new Bool[6];
+        using (x = Qubit[6])
+        {
+            ApplyToEach(H, x); // prepare the uniform distribution
+            for ((tz, tx) in Zip(timeZ, timeX))
+            {
+                InstanceHamiltonian(x, tz, weights, couplings); // do Exp(-i H_C tz)
+                DriverHamiltonian(x, tx); // do Exp(-i H_0 tx)
+            }
+            set result = ResultArrayAsBoolArray(MultiM(x)); // measure in the computational basis
+            ResetAll(x);
+        }
+        return result;
+    }
+
+    /// # Summary
+    /// Calculate the total cost for the given result.
+    ///
+    /// # Input
+    /// ## segmentCosts
+    /// Array of costs per segment
+    /// ## usedSegments
+    /// Array of which segments are used
+    ///
+    /// # Output
+    /// ## finalCost
+    /// Calculated cost of given path
+    function calculateCost(segmentCosts : Double[], usedSegments : Bool[]) : Double {
+        mutable finalCost = 0.0;
+        for ((cost, segment) in Zip(segmentCosts, usedSegments)) {
+            set finalCost = segment ? finalCost + cost | finalCost;
+        }
+        return finalCost;
+    }
+
+    /// # Summary
+    /// Final check to determine if the used segments satisfy our known constraints.
+    /// Returns true or false.
+    ///
+    /// # Input
+    /// ## usedSegments
+    /// Array of which segments were used
+    function determineSatisfactory(usedSegments : Bool[]) : Bool {
+        mutable HammingWeight = 0;
```

Comment:
This is currently named as a global callable (PascalCase) rather than a local variable (camelCase).